utils/crystal/memoization.cr

```diff
@@ -0,0 +1,102 @@
+module Memoization
+  macro memoize(method_def)
+    {% raise "You must define a return type for memoized methods" if method_def.return_type.is_a?(Nop) %}
+    {% raise "All arguments must have an explicit type restriction for memoized methods" if method_def.args.any? &.restriction.is_a?(Nop) %}
+
+    {% special_ending = nil %}
+    {% safe_method_name = method_def.name %}
+
+    {%
+      if method_def.name.ends_with?('?')
+        special_ending = "?"
+        safe_method_name = method_def.name.tr("?", "")
+      elsif method_def.name.ends_with?('!')
+        special_ending = "!"
+        safe_method_name = method_def.name.tr("!", "")
+      end
+    %}
+
+    {% if method_def.args.empty? %}
+      @__memoized_{{safe_method_name}} : {{ method_def.return_type }} | UninitializedMemo = UninitializedMemo::INSTANCE
+    {% else %}
+      @__memoized_{{safe_method_name}} : Hash(
+        Tuple(
+          {% for arg in method_def.args %}
+            {{ arg.restriction }},
+          {% end %}
+        ), {{ method_def.return_type }}
+      ) = Hash(
+        Tuple(
+          {% for arg in method_def.args %}
+            {{ arg.restriction }},
+          {% end %}
+        ), {{ method_def.return_type }}).new
```

Comment:
Here is mostly where I made a change to handle an indefinite number of memoizations for different method arguments.